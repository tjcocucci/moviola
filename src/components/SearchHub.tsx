import DatePicker from './DatePicker';
import SearchBar from './SearchBar';
import { useState } from 'react';
import Selector from './Selector';
import SearchParams from '@/types/SearchParams';

const defaultSearchParams: SearchParams = {
  query: '',
  startDate: null,
  endDate: null,
  sortBy: '',
};

export default function SearchHub({
  children,
  handleSearch,
}: {
  children: React.ReactNode;
  handleSearch: Function;
}) {
  const [searchParams, setSearchParams] = useState(defaultSearchParams);

  function handleTextQueryChange(textQuery: string) {
    const newParams = { ...searchParams, query: textQuery };
    setSearchParams(newParams);
    handleSearch(newParams);
  }

  async function handleStartDateChange(date: Date) {
    const newParams = { ...searchParams, startDate: date };
    setSearchParams(newParams);
    handleSearch(newParams);
  }

  async function handleEndDateChange(date: Date) {
    const newParams = { ...searchParams, endDate: date };
    setSearchParams(newParams);
    handleSearch(newParams);
  }

  function handleSortByChange(sortBy: string) {
    const newParams = { ...searchParams, sortBy: sortBy };
    setSearchParams(newParams);
    handleSearch(newParams);
  }

  return (
    <>
      <div className="p-4 flex flex-wrap">
        <div className="flex items-center mb-4">
          <SearchBar handleChange={handleTextQueryChange} />
        </div>
        <div className="flex items-center">
          <span className="mr-2">Start date</span>{' '}
          <DatePicker handleChange={handleStartDateChange} />
        </div>
        <div className="flex items-center">
          <span className="mr-2">End date</span>{' '}
          <DatePicker handleChange={handleEndDateChange} />
        </div>
        <div className="flex items-center">
          <span className="mr-2">Order by</span>
          <Selector handleChange={handleSortByChange} />
        </div>
      </div>
      {children}
    </>
  );
}
