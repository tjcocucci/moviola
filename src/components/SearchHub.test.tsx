import { describe, it, expect, vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';
import SearchHub from './SearchHub';

const defaultProps = {
  children: <div>Test</div>,
  handleSearch: () => {},
};

describe('[ components / SearchHub ]', () => {
  describe('when component is rendered', () => {
    it('should render snapshoted SearchHub', () => {
      const { asFragment } = render(<SearchHub {...defaultProps} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when search components are not changed', () => {
    it('should not call handleChange', () => {
      const props = {
        handleChange: vi.fn(),
      };
      expect(props.handleChange).not.toHaveBeenCalled();
    });
  });

  describe('when text is imputed in searchbar', () => {
    it('should call handleChange', () => {
      vi.useFakeTimers();
      const props = {
        children: <div>Test</div>,
        handleSearch: vi.fn(),
      };
      const { container } = render(<SearchHub {...props} />);
      const inputElement = container.querySelectorAll('input')[0];

      if (inputElement !== null) {
        fireEvent.change(inputElement, {
          target: { value: 'godfather' },
        });
      } else {
        throw new Error('Input element not found');
      }
      vi.advanceTimersByTime(501);

      expect(props.handleSearch).toHaveBeenCalled();
    });
  });

  describe('when start date is changed', () => {
    it('should call handleChange', () => {
      const props = {
        children: <div>Test</div>,
        handleSearch: vi.fn(),
      };
      const { container } = render(<SearchHub {...props} />);
      const inputElement = container.querySelectorAll('input')[1];

      if (inputElement !== null) {
        fireEvent.change(inputElement, {
          target: { value: '2020-01-01' },
        });
      } else {
        throw new Error('Input element not found');
      }
      expect(props.handleSearch).toHaveBeenCalled();
    });
  });

  describe('when end date is changed', () => {
    it('should call handleChange', () => {
      const props = {
        children: <div>Test</div>,
        handleSearch: vi.fn(),
      };
      const { container } = render(<SearchHub {...props} />);
      const inputElement = container.querySelectorAll('input')[2];

      if (inputElement !== null) {
        fireEvent.change(inputElement, {
          target: { value: '2020-01-01' },
        });
      } else {
        throw new Error('Input element not found');
      }
      expect(props.handleSearch).toHaveBeenCalled();
    });
  });

  describe('when option is changed', () => {
    it('should call handleChange', () => {
      const props = {
        children: <div>Test</div>,
        handleSearch: vi.fn(),
      };
      const { container } = render(<SearchHub {...props} />);
      const selectElement = container.querySelectorAll('select')[0];

      if (selectElement !== null) {
        fireEvent.change(selectElement, {
          target: { value: 'popularity.desc' },
        });
      } else {
        throw new Error('select element not found');
      }
      expect(props.handleSearch).toHaveBeenCalled();
    });
  });
});
