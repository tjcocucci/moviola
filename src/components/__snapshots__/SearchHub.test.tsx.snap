// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[ components / SearchHub ] > when component is rendered > should render snapshoted SearchHub 1`] = `
<DocumentFragment>
  <div>
    <div
      class="p-4 flex self-center flex-auto flex-col gap-5 max-w-4xl"
    >
      <div
        class="mb-3"
      >
        <input
          class="p-2 border border-gray-300 rounded mr-2"
          placeholder="Search..."
          type="text"
          value=""
        />
      </div>
      <div
        class="flex"
      >
        <div
          class="flex items-center"
        >
          <span
            class="whitespace-nowrap mr-5"
          >
            Start date:
          </span>
          <div
            class="react-datepicker-wrapper"
          >
            <div
              class="react-datepicker__input-container "
            >
              <span
                aria-live="polite"
                class="react-datepicker__aria-live"
                role="alert"
              />
              <input
                class="p-2 border border-gray-300 rounded w-2/3"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="flex items-center"
        >
          <span
            class="whitespace-nowrap mr-5"
          >
            End date:
          </span>
          <div
            class="react-datepicker-wrapper"
          >
            <div
              class="react-datepicker__input-container "
            >
              <span
                aria-live="polite"
                class="react-datepicker__aria-live"
                role="alert"
              />
              <input
                class="p-2 border border-gray-300 rounded w-2/3"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="flex items-center "
        >
          <div
            class="p-2 "
          >
            <select>
              <option
                value=""
              >
                Order by
              </option>
              <option
                value="popularity.desc"
              >
                Popularity
              </option>
              <option
                value="release_date.desc"
              >
                Release date
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
    <div>
      Test
    </div>
  </div>
</DocumentFragment>
`;
